File: swift/monosecret_swift/Sources/Monosecret/Monosecret.swift
import Foundation

private let resolveSchemaVersion = 2
private let reportSchemaVersion = 1

/// Caller-asserted software-integration context (Monosecret 0.20+).
public struct CallerContext: Encodable, Sendable {
    public let name: String
    public let version: String?
    public let operation: String?
    public let resource: String?

    public init(
        name: String,
        version: String? = nil,
        operation: String? = nil,
        resource: String? = nil
    ) {
        self.name = name
        self.version = version
        self.operation = operation
        self.resource = resource
    }
}

private struct ResolveRequest: Encodable, Sendable {
    var path: String? = nil
    var provider: String? = nil
    var profile: String? = nil
    var scope: String? = nil
    var reason: String? = nil
    var caller: CallerContext? = nil
    var noValues: Bool? = nil
    var mode: String? = nil

    enum CodingKeys: String, CodingKey {
        case path
        case provider
        case profile
        case scope
        case reason
        case caller
        case noValues = "no_values"
        case mode
    }
}

private struct ErrorPayload: Decodable {
    let kind: String?
    let message: String?
}

private struct Envelope<Response: Decodable>: Decodable {
    let ok: Bool
    let response: Response?
    let error: ErrorPayload?
}

private struct ResolveResponse: Decodable {
    let schemaVersion: Int
    let provider: String
    let profile: String
    let scope: String?
    let secrets: [String: ResolvedSecret]
    let missingRequired: [String]
    let missingOptional: [String]

    enum CodingKeys: String, CodingKey {
        case schemaVersion = "schema_version"
        case provider
        case profile
        case scope
        case secrets
        case missingRequired = "missing_required"
        case missingOptional = "missing_optional"
    }
}

private struct ReportResponse: Decodable {
    let schemaVersion: Int
    let provider: String
    let profile: String
    let scope: String?
    let secrets: [SecretReport]
    let constraintViolations: [ConstraintViolation]?

    enum CodingKeys: String, CodingKey {
        case schemaVersion = "schema_version"
        case provider
        case profile
        case scope
        case secrets
        case constraintViolations = "constraint_violations"
    }
}

/// Configures a Monosecret resolution.
public struct MonosecretBuilder: Sendable {
    private var request = ResolveRequest()
    private var inline: InlineSpec? = nil

    private struct InlineSpec: Sendable {
        let declaration: Data
        let baseDir: String
    }

    public init() {}

    public func withPath(_ path: String?) -> Self {
        var copy = setting(\.path, to: path)
        copy.inline = nil
        return copy
    }

    /// Resolves strict inline-spec v1 at `baseDir` (Monosecret 0.20+).
    ///
    /// The declaration is encoded once into the dedicated native wire format.
    /// An older native library reports a capability error rather than searching
    /// for a filesystem manifest.
    public func withInlineSpec<Declaration: Encodable & Sendable>(
        _ declaration: Declaration,
        baseDir: String
    ) throws -> Self {
        var copy = self
        copy.request.path = nil
        copy.inline = InlineSpec(
            declaration: try JSONEncoder().encode(declaration), baseDir: baseDir
        )
        return copy
    }

    public func withProvider(_ provider: String?) -> Self {
        setting(\.provider, to: provider)
    }

    public func withProfile(_ profile: String?) -> Self {
        setting(\.profile, to: profile)
    }

    /// Limits resolution to a named manifest scope.
    public func withScope(_ scope: String?) -> Self {
        setting(\.scope, to: scope)
    }

    public func withReason(_ reason: String?) -> Self {
        setting(\.reason, to: reason)
    }

    /// Identifies the invoking software integration (Monosecret 0.20+).
    public func withCaller(_ caller: CallerContext?) -> Self {
        setting(\.caller, to: caller)
    }

    public func withNoValues(_ noValues: Bool = true) -> Self {
        setting(\.noValues, to: noValues)
    }

    /// Resolves the configured secrets.
    ///
    /// - Throws: ``MissingRequiredError`` when required secrets are absent, or
    ///   ``MonosecretError`` for any other failure.
    public func load() throws -> Resolved {
        let response: ResolveResponse = try call(mode: nil)
        try ensureSchemaVersion(
            actual: response.schemaVersion,
            expected: resolveSchemaVersion,
            kind: "resolve"
        )
        if !response.missingRequired.isEmpty {
            throw MissingRequiredError(missing: response.missingRequired)
        }

        return Resolved(
            provider: response.provider,
            profile: response.profile,
            scope: response.scope,
            secrets: response.secrets,
            missingOptional: response.missingOptional
        )
    }

    /// Builds a value-free inventory report.
    ///
    /// Missing required secrets appear in the report rather than throwing.
    public func report() throws -> ResolutionReport {
        let response: ReportResponse = try call(mode: "report")
        try ensureSchemaVersion(
            actual: response.schemaVersion,
            expected: reportSchemaVersion,
            kind: "report"
        )
        return ResolutionReport(
            provider: response.provider,
            profile: response.profile,
            scope: response.scope,
            secrets: response.secrets,
            constraintViolations: response.constraintViolations ?? []
        )
    }

    private func setting<Value>(
        _ keyPath: WritableKeyPath<ResolveRequest, Value>,
        to value: Value
    ) -> Self {
        var copy = self
        copy.request[keyPath: keyPath] = value
        return copy
    }

    private func call<Response: Decodable>(mode: String?) throws -> Response {
        var configured = request
        configured.mode = mode

        let requestData: Data
        do {
            if let inline {
                let declaration = try JSONSerialization.jsonObject(with: inline.declaration)
                requestData = try JSONSerialization.data(withJSONObject: [
                    "request_version": 1,
                    "operation": "resolve",
                    "source": [
                        "kind": "inline", "spec_version": 1,
                        "base_dir": inline.baseDir, "spec": declaration,
                    ],
                    "options": try JSONSerialization.jsonObject(with: JSONEncoder().encode(configured)),
                ])
            } else {
                requestData = try JSONEncoder().encode(configured)
            }
        } catch {
            throw MonosecretError(kind: "encode", message: error.localizedDescription)
        }
        guard let requestJSON = String(data: requestData, encoding: .utf8) else {
            throw MonosecretError(
                kind: "encode",
                message: "could not encode the request as UTF-8"
            )
        }

        let responseJSON: String
        if inline == nil {
            responseJSON = try Native.resolve(requestJSON)
        } else {
            responseJSON = try Native.call(requestJSON)
        }
        let envelope: Envelope<Response>
        do {
            envelope = try JSONDecoder().decode(
                Envelope<Response>.self,
                from: Data(responseJSON.utf8)
            )
        } catch {
            throw MonosecretError(kind: "parse", message: error.localizedDescription)
        }

        guard envelope.ok else {
            throw MonosecretError(
                kind: envelope.error?.kind ?? "unknown",
                message: envelope.error?.message
                    ?? "native resolver returned an unspecified error"
            )
        }
        guard let response = envelope.response else {
            throw MonosecretError(
                kind: "ffi",
                message: "monosecret_resolve reported ok with no response"
            )
        }
        return response
    }

    private func ensureSchemaVersion(
        actual: Int,
        expected: Int,
        kind: String
    ) throws {
        guard actual == expected else {
            throw MonosecretError(
                kind: "version",
                message: "unsupported \(kind) schema version \(actual) "
                    + "(expected \(expected)); the libmonosecret_ffi library "
                    + "and this SDK are out of sync"
            )
        }
    }
}

/// Entry point for the Monosecret Swift SDK.
public enum Monosecret {
    public static func builder() -> MonosecretBuilder {
        MonosecretBuilder()
    }

    /// Resolves secrets in one call.
    public static func resolve(
        path: String? = nil,
        provider: String? = nil,
        profile: String? = nil,
        scope: String? = nil,
        reason: String? = nil,
        caller: CallerContext? = nil
    ) throws -> Resolved {
        try configured(
            path: path,
            provider: provider,
            profile: profile,
            scope: scope,
            reason: reason,
            caller: caller
        ).load()
    }

    /// Builds a value-free inventory report in one call.
    public static func report(
        path: String? = nil,
        provider: String? = nil,
        profile: String? = nil,
        scope: String? = nil,
        reason: String? = nil,
        caller: CallerContext? = nil
    ) throws -> ResolutionReport {
        try configured(
            path: path,
            provider: provider,
            profile: profile,
            scope: scope,
            reason: reason,
            caller: caller
        ).report()
    }

    /// The ABI version reported by the bundled native resolver.
    public static func abiVersion() throws -> String {
        try Native.abiVersion()
    }

    private static func configured(
        path: String?,
        provider: String?,
        profile: String?,
        scope: String?,
        reason: String?,
        caller: CallerContext?
    ) -> MonosecretBuilder {
        builder()
            .withPath(path)
            .withProvider(provider)
            .withProfile(profile)
            .withScope(scope)
            .withReason(reason)
            .withCaller(caller)
    }
}

File: go/monosecret_go/monosecret.go
// Package monosecret is a Go SDK for Monosecret, a declarative secrets manager.
//
// It is a thin client over the monosecret_ffi C ABI. Resolution (providers,
// chains, profiles, generation, as_path) happens entirely in the Rust core; this
// package marshals a JSON request to monosecret_resolve, parses the response
// envelope, and exposes it with the same vocabulary as the Rust derive crate.
//
// Three build modes select the native resolver:
//   - default (no build tag): purego (dlopen, no cgo). The library is located via
//     the MONOSECRET_FFI_LIB environment variable, an embedded copy, or a Cargo
//     target directory. This keeps `go get` toolchain-free.
//   - `-tags monosecret_static`: cgo statically links libmonosecret_ffi.a, so the resolver
//     is embedded in the Go binary (fully static on Linux/musl).
//   - `-tags pkgconfig` (0.19+): cgo links the installed static or shared library
//     described by monosecret_ffi.pc. See README.
//
// All bindings implement the same hooks (ensureLoaded, nativeResolve,
// nativeABIVersion); the code below is binding-agnostic.
package monosecret

import (
	"encoding/json"
	"fmt"
	"os"
	"strings"
)

// resolveSchemaVersion is the response wire-format version this SDK understands.
// It tracks monosecret_ffi's RESOLVE_SCHEMA_VERSION; a mismatch means the loaded
// library is incompatible with this SDK, so Load reports it rather than silently
// misparsing.
const resolveSchemaVersion = 2

// Error is a resolution failure (bad manifest, provider error, reason policy).
type Error struct {
	Kind    string
	Message string
}

func (e *Error) Error() string {
	return fmt.Sprintf("%s (kind: %s)", e.Message, e.Kind)
}

// MissingRequiredError reports required secrets that were not found anywhere.
type MissingRequiredError struct {
	Missing []string
}

// CallerContext identifies the software integration invoking Monosecret.
// It is caller-asserted audit metadata and never supplies an access reason.
// Available since Monosecret 0.20.
type CallerContext struct {
	Name      string `json:"name"`
	Version   string `json:"version,omitempty"`
	Operation string `json:"operation,omitempty"`
	Resource  string `json:"resource,omitempty"`
}

func (e *MissingRequiredError) Error() string {
	return "missing required secret(s): " + strings.Join(e.Missing, ", ")
}

// ResolvedSecret is one resolved secret. Exactly one of Value / Path is set.
type ResolvedSecret struct {
	Value          *string
	Path           *string
	AsPath         bool
	Source         string
	SourceProvider *string
}

// Usable returns the secret's usable string and whether one is present: the file
// path for as_path secrets, otherwise the value. Both are absent when a
// value-less response (e.g. no_values) strips them, in which case ok is false.
// This is the null-aware accessor; the other SDKs express the same thing as a
// get() that returns null/None/nil.
func (s ResolvedSecret) Usable() (string, bool) {
	if s.AsPath {
		if s.Path != nil {
			return *s.Path, true
		}
		return "", false
	}
	if s.Value != nil {
		return *s.Value, true
	}
	return "", false
}

// Get returns the usable string: the file path for as_path secrets, else the
// value. It is the empty string when no usable value is present; use Usable to
// distinguish an absent value from a genuinely empty one.
func (s ResolvedSecret) Get() string {
	v, _ := s.Usable()
	return v
}

// Resolved is a successful resolution, mirroring the Rust Resolved wrapper.
type Resolved struct {
	Provider string
	Profile  string
	// Scope is the selected manifest scope, or nil for a full-profile resolve (0.17+).
	Scope           *string
	Secrets         map[string]ResolvedSecret
	MissingOptional []string
}

// SetAsEnv exports each resolved secret into the process environment by name.
// Secrets with no usable value (e.g. under no_values) are skipped rather than
// exported as an empty string.
func (r *Resolved) SetAsEnv() error {
	for name, secret := range r.Secrets {
		if value, ok := secret.Usable(); ok {
			if err := os.Setenv(name, value); err != nil {
				return err
			}
		}
	}
	return nil
}

// Fields returns a flat map of SECRET_NAME -> value (the file path for as_path).
// A secret with no usable value (e.g. under no_values) maps to a nil pointer,
// which marshals to JSON null, matching the null the Python, Ruby, and Node SDKs
// emit; the value is a non-nil pointer otherwise.
func (r *Resolved) Fields() map[string]*string {
	out := make(map[string]*string, len(r.Secrets))
	for name, secret := range r.Secrets {
		if v, ok := secret.Usable(); ok {
			val := v
			out[name] = &val
		} else {
			out[name] = nil
		}
	}
	return out
}

// FieldsJSON marshals Fields() to JSON (a `{SECRET_NAME: value-or-null}` object),
// the input for a quicktype-generated deserializer (e.g. UnmarshalMonosecret).
// See `monosecret schema`.
func (r *Resolved) FieldsJSON() ([]byte, error) {
	return json.Marshal(r.Fields())
}

// Close removes the temp files backing any as_path secrets in this result. The
// resolver persists those files (mode 0400) so their paths stay valid after
// resolve returns; the caller owns their lifetime. Call it (e.g.
// `defer resolved.Close()`) when done so secret files do not accumulate in the
// temp dir. Non-as_path secrets and a no_values result hold no path and are
// skipped, and a file already gone is not an error.
func (r *Resolved) Close() error {
	var firstErr error
	for _, secret := range r.Secrets {
		if secret.AsPath && secret.Path != nil {
			if err := os.Remove(*secret.Path); err != nil && !os.IsNotExist(err) && firstErr == nil {
				firstErr = err
			}
		}
	}
	return firstErr
}

// ABIVersion returns the version reported by the native resolver.
func ABIVersion() (string, error) {
	if err := ensureLoaded(); err != nil {
		return "", err
	}
	return nativeABIVersion()
}

// Builder configures a resolution, mirroring the derive crate's Monosecret::builder().
type Builder struct {
	req    map[string]any
	inline *inlineSource
}

type inlineSource struct {
	baseDir string
	spec    any
}

// New starts a resolution builder.
func New() *Builder {
	return &Builder{req: map[string]any{}}
}

// set lazily initializes the request map so a zero-value Builder (e.g.
// `var b Builder` or `&Builder{}`, not just New()) does not panic with a
// nil-map write in the setters below.
func (b *Builder) set(key string, value any) *Builder {
	if b.req == nil {
		b.req = map[string]any{}
	}
	b.req[key] = value
	return b
}

func (b *Builder) WithPath(path string) *Builder {
	// Sources are a tagged union in the native request. Choosing a path after
	// an inline source deliberately selects the legacy path source instead of
	// serializing an ambiguous request.
	b.inline = nil
	return b.set("path", path)
}
func (b *Builder) WithProvider(p string) *Builder { return b.set("provider", p) }
func (b *Builder) WithProfile(p string) *Builder  { return b.set("profile", p) }

// WithScope limits resolution to a named manifest scope (Monosecret 0.17+).
func (b *Builder) WithScope(scope string) *Builder   { return b.set("scope", scope) }
func (b *Builder) WithReason(reason string) *Builder { return b.set("reason", reason) }
func (b *Builder) WithCaller(caller CallerContext) *Builder {
	return b.set("caller", caller)
}
func (b *Builder) WithNoValues(v bool) *Builder { return b.set("no_values", v) }

// WithInlineSpec resolves a strict, versioned inline declaration instead of a
// filesystem manifest. `spec` is serialized as the native inline-spec v1
// document (project/profiles/secrets); `baseDir` resolves relative provider
// paths just like the directory of a manifest. Available since Monosecret 0.20.
//
// The native library must export `monosecret_call`. If it is older, Load and
// Report return a capability error rather than falling back to a manifest search.
func (b *Builder) WithInlineSpec(spec any, baseDir string) *Builder {
	if b.req != nil {
		delete(b.req, "path")
	}
	b.inline = &inlineSource{baseDir: baseDir, spec: spec}
	return b
}

// envelope is the response wrapper shared by the resolve and report paths,
// generic over its inner response type R.
type envelope[R any] struct {
	OK       bool       `json:"ok"`
	Response *R         `json:"response"`
	Error    *errorJSON `json:"error"`
}

type errorJSON struct {
	Kind    string `json:"kind"`
	Message string `json:"message"`
}

type secretJSON struct {
	Value          *string `json:"value"`
	Path           *string `json:"path"`
	AsPath         bool    `json:"as_path"`
	Source         string  `json:"source"`
	SourceProvider *string `json:"source_provider"`
}

type responseJSON struct {
	SchemaVersion   int                   `json:"schema_version"`
	Provider        string                `json:"provider"`
	Profile         string                `json:"profile"`
	Scope           *string               `json:"scope"`
	Secrets         map[string]secretJSON `json:"secrets"`
	MissingRequired []string              `json:"missing_required"`
	MissingOptional []string              `json:"missing_optional"`
}

// parseEnvelope unmarshals a response envelope, validates the envelope-level
// fields shared by the resolve and report responses, and returns the inner
// response. kind ("resolve" or "report") labels the version-mismatch message;
// schemaOf reads the response's schema_version for the version check.
func parseEnvelope[R any](raw, kind string, expected int, schemaOf func(*R) int) (*R, error) {
	var env envelope[R]
	if err := json.Unmarshal([]byte(raw), &env); err != nil {
		return nil, err
	}
	if !env.OK {
		errKind, message := "unknown", ""
		if env.Error != nil {
			errKind, message = env.Error.Kind, env.Error.Message
		}
		return nil, &Error{Kind: errKind, Message: message}
	}
	if env.Response == nil {
		return nil, &Error{Kind: "ffi", Message: "monosecret_resolve reported ok with no response"}
	}
	if v := schemaOf(env.Response); v != expected {
		return nil, &Error{Kind: "version", Message: fmt.Sprintf(
			"unsupported %s schema version %d (expected %d); the monosecret_ffi library and this SDK are out of sync",
			kind, v, expected,
		)}
	}
	return env.Response, nil
}

// Load resolves the secrets. It returns *MissingRequiredError if a required
// secret is missing, and *Error for any other failure.
func (b *Builder) Load() (*Resolved, error) {
	// A zero-value Builder (var b Builder; b.Load()) has a nil req, which marshals
	// to the literal `null` that serde rejects as an invalid JsonRequest. The WithX
	// setters lazily allocate req, but Load may run before any of them.
	if b.req == nil {
		b.req = map[string]any{}
	}
	raw, err := b.execute("")
	if err != nil {
		return nil, err
	}

	resp, err := parseEnvelope(raw, "resolve", resolveSchemaVersion, func(r *responseJSON) int { return r.SchemaVersion })
	if err != nil {
		return nil, err
	}
	if len(resp.MissingRequired) > 0 {
		return nil, &MissingRequiredError{Missing: resp.MissingRequired}
	}

	secrets := make(map[string]ResolvedSecret, len(resp.Secrets))
	for name, entry := range resp.Secrets {
		secrets[name] = ResolvedSecret{
			Value:          entry.Value,
			Path:           entry.Path,
			AsPath:         entry.AsPath,
			Source:         entry.Source,
			SourceProvider: entry.SourceProvider,
		}
	}
	return &Resolved{
		Provider:        resp.Provider,
		Profile:         resp.Profile,
		Scope:           resp.Scope,
		Secrets:         secrets,
		MissingOptional: resp.MissingOptional,
	}, nil
}

// reportSchemaVersion is the value-free report wire-format version this SDK
// understands; it tracks monosecret's RESOLUTION_REPORT_SCHEMA_VERSION.
const reportSchemaVersion = 1

// SecretReport is the value-free resolution outcome for one declared secret:
// how it would resolve and from where, never the value itself.
type SecretReport struct {
	Name           string
	Status         string // "resolved" | "missing_required" | "missing_optional"
	Required       bool
	SourceProvider *string
	DefaultApplied bool
	Generated      bool
	AsPath         bool
}

// ConstraintViolationKind identifies a cross-secret presence constraint.
type ConstraintViolationKind string

const (
	ConstraintViolationAtLeastOne ConstraintViolationKind = "at_least_one"
	ConstraintViolationExactlyOne ConstraintViolationKind = "exactly_one"
)

// ConstraintViolation describes a failed cross-secret presence constraint.
type ConstraintViolation struct {
	Kind    ConstraintViolationKind `json:"kind"`
	Group   string                  `json:"group"`
	Secrets []string                `json:"secrets"`
	Present []string                `json:"present"`
}

// Report is a value-free resolution snapshot: every declared secret and how it
// would resolve, never a value. Unlike Load, a missing required secret is
// reported as a SecretReport with Status "missing_required" rather than an
// error, so it describes a profile even when its secrets are not all available.
type Report struct {
	Provider string
	Profile  string
	// Scope is the selected manifest scope, or nil for a full-profile report (0.17+).
	Scope                *string
	Secrets              []SecretReport
	ConstraintViolations []ConstraintViolation
}

type secretReportJSON struct {
	Name           string  `json:"name"`
	Status         string  `json:"status"`
	Required       bool    `json:"required"`
	SourceProvider *string `json:"source_provider"`
	DefaultApplied bool    `json:"default_applied"`
	Generated      bool    `json:"generated"`
	AsPath         bool    `json:"as_path"`
}

type reportResponseJSON struct {
	SchemaVersion        int                   `json:"schema_version"`
	Provider             string                `json:"provider"`
	Profile              string                `json:"profile"`
	Scope                *string               `json:"scope"`
	Secrets              []secretReportJSON    `json:"secrets"`
	ConstraintViolations []ConstraintViolation `json:"constraint_violations"`
}

// Report resolves the value-free report (the inventory/preflight view, the same
// one the CLI exposes as `check --json`). It never returns
// *MissingRequiredError: a missing required secret appears as a SecretReport
// with Status "missing_required". It returns *Error for a genuine failure.
func (b *Builder) Report() (*Report, error) {
	raw, err := b.execute("report")
	if err != nil {
		return nil, err
	}

	resp, err := parseEnvelope(raw, "report", reportSchemaVersion, func(r *reportResponseJSON) int { return r.SchemaVersion })
	if err != nil {
		return nil, err
	}

	secrets := make([]SecretReport, len(resp.Secrets))
	for i, s := range resp.Secrets {
		secrets[i] = SecretReport(s)
	}
	constraintViolations := resp.ConstraintViolations
	if constraintViolations == nil {
		constraintViolations = []ConstraintViolation{}
	}
	return &Report{
		Provider:             resp.Provider,
		Profile:              resp.Profile,
		Scope:                resp.Scope,
		Secrets:              secrets,
		ConstraintViolations: constraintViolations,
	}, nil
}

// execute keeps the legacy request exactly as it was for path/search callers.
// Inline callers use the separate C symbol, making an old runtime's missing
// capability an explicit load error instead of a filesystem fallback.
func (b *Builder) execute(mode string) (string, error) {
	if b.req == nil {
		b.req = map[string]any{}
	}
	options := make(map[string]any, len(b.req)+1)
	for k, v := range b.req {
		options[k] = v
	}
	if mode != "" {
		options["mode"] = mode
	}
	if b.inline == nil {
		if err := ensureLoaded(); err != nil {
			return "", err
		}
		payload, err := json.Marshal(options)
		if err != nil {
			return "", err
		}
		return nativeResolve(string(payload))
	}
	if err := ensureCallLoaded(); err != nil {
		return "", err
	}
	payload, err := json.Marshal(map[string]any{
		"request_version": 1,
		"operation":       "resolve",
		"source": map[string]any{
			"kind": "inline", "spec_version": 2,
			"base_dir": b.inline.baseDir, "spec": b.inline.spec,
		},
		"options": options,
	})
	if err != nil {
		return "", err
	}
	return nativeCall(string(payload))
}
